gnomad/sample_qc/filtering.py

```diff
@@ -362,3 +442,196 @@ def merge_sample_qc_expr(
+                group_n,
+            )
+        if use_approximation:
+            nbrs = AnnoyIndex(n_pcs, distance_metric)
+            for i, row in scores_pd.iterrows():
+                nbrs.add_item(i, row)
+            nbrs.build(n_trees, n_jobs=n_jobs)
+
+            indexes = []
+            for i in range(group_n):
+                indexes.append(
+                    nbrs.get_nns_by_item(
+                        i, n_neighbors, include_distances=add_neighbor_distances
+                    )
+                )
+            if add_neighbor_distances:
+                distances = [d for i, d in indexes]
+                indexes = [i for i, d in indexes]
+        else:
+            scores = scores_pd.values
+            nbrs = NearestNeighbors(
+                n_neighbors=n_neighbors, n_jobs=n_jobs, metric=distance_metric
+            )
+            nbrs.fit(scores)
+            indexes = nbrs.kneighbors(scores, return_distance=add_neighbor_distances)
+            if add_neighbor_distances:
+                distances, indexes = indexes
+
+        # Format neighbor indexes as a Hail Table.
+        indexes_pd = pd.DataFrame(indexes)
+        indexes_pd = pd.concat([scores_pd_s, indexes_pd], axis=1)
+        indexes_pd = indexes_pd.rename(
+            columns={i: f"nbrs_index_{i}" for i in range(n_neighbors)}
+        )
+        indexes_ht = hl.Table.from_spark(spark.createDataFrame(indexes_pd), key=["s"])
+        indexes_ht = indexes_ht.transmute(
+            nearest_neighbor_idxs=hl.array(
+                [indexes_ht[f"nbrs_index_{i}"] for i in range(n_neighbors)]
+            )
+        )
+
+        if add_neighbor_distances:
+            # Format neighbor distances as a Hail Table.
+            distances_pd = pd.DataFrame(distances)
+            distances_pd = distances_pd.rename(
+                columns={i: f"nbrs_{i}" for i in range(n_neighbors)}
+            )
+            distances_pd = pd.concat([scores_pd_s, distances_pd], axis=1)
+            distances_ht = hl.Table.from_spark(
+                spark.createDataFrame(distances_pd), key=["s"]
+            )
+            distances_ht = distances_ht.transmute(
+                nearest_neighbor_dists=hl.array(
+                    [distances_ht[f"nbrs_{str(i)}"] for i in range(n_neighbors)]
+                )
+            )
+
+            # Join neighbor distances Table and indexes Table.
+            nbrs_ht = indexes_ht.join(distances_ht)
+        else:
+            nbrs_ht = indexes_ht
+
+        # Add nearest_neighbors annotation to use instead of indexes.
+        nbrs_ht = nbrs_ht.add_index()
+        explode_nbrs_ht = nbrs_ht.key_by("idx").explode("nearest_neighbor_idxs")
+        explode_nbrs_ht = explode_nbrs_ht.transmute(
+            nbr=explode_nbrs_ht[hl.int64(explode_nbrs_ht.nearest_neighbor_idxs)].s
+        )
+        explode_nbrs_ht = explode_nbrs_ht.group_by("s").aggregate(
+            nearest_neighbors=hl.agg.collect_as_set(explode_nbrs_ht.nbr)
```

Comment:
does maintaining the order when collecting matter?

Reply:
Oh good point, if we ever want to map to the `nearest_neighbor_dists` array it needs to be in the same order